### PR TITLE
feat(Harness): support user secrets

### PR DESCRIPTION
**Pull request by IronCore864**
Add support for user secrets in Harness. Changes:

- Add a new API `Harness.add_user_secret` which simulates `juju add-secret`.
- `add_model_secret` is deprecated, use `add_charm_secret`. For now, `add_model_secret` is an alias to `add_charm_secret`.

Use case example:

`config.yaml`:

```yaml
options:
  mysec:
    type: secret
    description: "tell me your secrets"
```

`charm.py`:

```python
class MyVMCharm(ops.CharmBase):
    def __init__(self, framework: ops.Framework):
        super().__init__(framework)
        framework.observe(self.on.config_changed, self._on_config_changed)
    def _on_config_changed(self, event: ops.StartEvent):
        mysec = self.config.get('mysec')
        if mysec:
            sec = self.model.get_secret(id=mysec, label="mysec")
            self.config_from_secret = sec.get_content()
```

`test_charm.py`:

```python
def test_config_changed(harness):
    secret_content = {'password': 'foo'}
    secret_id = harness.add_user_secret(secret_content)
    # grant the user secret to the app
    harness.grant_secret(secret_id, 'webapp')
    harness.begin()
    harness.update_config({'mysec': secret_id})
    secret = harness.model.get_secret(id=secret_id).get_content()
    assert harness.charm.config_from_secret == secret.get_content()
```

### CODE REVIEW

File: ops/testing.py
Comment (tonyandrewmeyer) [inline ops/testing.py:1530]: For what it's worth, I think if we do change the name, then we should have a warning emitted here. Not sure whether `DeprecationWarning` or `PendingDeprecationWarning`.

**Comment by IronCore864** [comment]
Get user secret grant method is removed according to discussion. Will create a separate PR after this one to mark `add_model_secret` as deprecated.

File: ops/testing.py
Comment (tonyandrewmeyer) [inline ops/testing.py:1558]: This feels a bit artificial - in reality you'd have a charm that did something with the secret, and since the charm can't make any changes to the secret you wouldn't ever be asserting on it afterwards.

However, I'm not sure how detailed we want these little examples in the API reference docs to be - a more realistic example is perhaps better suited to a how-to guide, so maybe leave this as-is?

**Review by tonyandrewmeyer — CHANGES_REQUESTED** [review]
Looking good - a few little nits and a couple of queries.

Other questions/comments:

* The `trigger_secret_expiration` method should probably raise an error if it's used with a user secret. Is that already the case via permissions checking?
* Likewise for `trigger_secret_rotation`.
* It doesn't seem like we are testing the permission code changes. I think we'd want to have tests that verify that those work as expected.

**Review by benhoyt — COMMENTED** [review]
I agree with Tony's comments. I'll review further later. For now, can you please add a PR description with a brief description of the new API and with a usage example?

File: ops/testing.py
Comment (IronCore864) [reply ops/testing.py:1611]: It's true that from Harness's standpoint, there is no need to grant it to the local app but since it's simulating `juju grant-secret`, technically it can grant to other apps. Should I keep it as a set? Or simply as a str of the app name?

Comment (tonyandrewmeyer) [reply ops/testing.py:1611]: My point is that Harness doesn't run anything from other apps, so it essentially has no effect. I think keeping it is ok though.

File: ops/testing.py
Comment (IronCore864) [reply ops/testing.py:1558]: Yeah, it feels totally made-up and has no real-world guidance value at all. Let me whip up a useful usecase.

**Comment by IronCore864** [comment]
> Looking good - a few little nits and a couple of queries.
> 
> Other questions/comments:
> 
> * The `trigger_secret_expiration` method should probably raise an error if it's used with a user secret. Is that already the case via permissions checking?
> * Likewise for `trigger_secret_rotation`.
> * It doesn't seem like we are testing the permission code changes. I think we'd want to have tests that verify that those work as expected.

Changed `trigger_secret_expiration` and `trigger_secret_rotation` to raise an error on user secret, also added two unit tests.

For user secret permissions, there are some unit tests.

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:1756]:
```suggestion
            raise RuntimeError("Cannot trigger the secret-expired event for a user secret.")
```

File: ops/testing.py
Comment (IronCore864) [reply ops/testing.py:2650]: Touché. It can't be both, I'll remove the user_secret part.

**Review by benhoyt — APPROVED** [review]

File: ops/testing.py
Comment (tonyandrewmeyer) [inline ops/testing.py:1518]: Do we want to keep this comment (with the method name updated) in `add_charm_secret`? I think it's ok to remove it from the docstring, but I think it would be good to keep as a code comment?

File: ops/testing.py
Comment (tonyandrewmeyer) [inline ops/testing.py:1559]:
```suggestion
                def _on_config_changed(self, event: ops.ConfigChangedEvent):
```

File: ops/testing.py
Comment (tonyandrewmeyer) [inline ops/testing.py:1569]: I think we should have this as a working example (I'd like to get these doctest'd in the future). Presumably this would if you had a `charmcraft.yaml` where the name was `webapp` but by default it should be:

```suggestion
                harness.grant_secret(secret_id, 'test-charm')
```

(That's the name that Harness uses if there isn't one in the metadata).

File: ops/testing.py
Comment (tonyandrewmeyer) [inline ops/testing.py:1551]: We should use the more modern charmcraft approach.

```suggestion
            # charmcraft.yaml
            config:
              options:
                mysec:
                  type: secret
                  description: "tell me your secrets"
```

File: ops/testing.py
Comment (tonyandrewmeyer) [inline ops/testing.py:1577]:
```suggestion
        # Although it's named a user-owned secret in Juju, technically, the owner is the
```

File: ops/testing.py
Comment (tonyandrewmeyer) [inline ops/testing.py:1578]: I don't think it makes sense to use ReST for comments.

```suggestion
        # Model, so the secret's owner is set to `Model.uuid`.
```

**Review by tonyandrewmeyer — CHANGES_REQUESTED** [review]
Can we adjust `_ensure_secret_owner` so that it explicitly talks about user secrets? It could have a `user_secret = secret.owner_name == self.model.uuid` and an `and not user_secret`, although that isn't strictly necessary, so maybe just a comment explaining the user secret situation, with the comment that explains unit and app permissions?

I would also feel more comfortable if we had tests that explicitly made sure that user secrets fail ownership checks. This will prevent regressions in the future if we change how the permissions checks are done.

**Comment by IronCore864** [comment]
> Can we adjust `_ensure_secret_owner` so that it explicitly talks about user secrets? It could have a `user_secret = secret.owner_name == self.model.uuid` and an `and not user_secret`, although that isn't strictly necessary, so maybe just a comment explaining the user secret situation, with the comment that explains unit and app permissions?
> 
> I would also feel more comfortable if we had tests that explicitly made sure that user secrets fail ownership checks. This will prevent regressions in the future if we change how the permissions checks are done.

Yes it makes sense. I have added a test to try to do Secret.get_info to validate the ownership.

File: ops/testing.py
Comment (benhoyt) [reply ops/testing.py:1517]: I just talked to John Meinel about this. It sounds like they have *not* settled on the name "charm secrets" and there's a bit more discussion they want to have about all this. So I think we should wait with this renaming. We can either just hold off on merging this PR till they've settled that (my recommendation) or we could merge with `add_model_secret` and `add_user_secret` and then change/deprecated `add_model_secret` in a future PR. What do you think @tonyandrewmeyer?

Comment (tonyandrewmeyer) [reply ops/testing.py:1517]: Did John have any idea of when they might settle? If the expectation is that it'll be before 2.13 then I would definitely just wait. If it's more likely to be in May (e.g. after Madrid) then I'd leave it to @IronCore864 - if he wants to do the extra work to split this into two PRs I have no objection, but I also don't see any issue with waiting a bit longer. If it might be even further away than that, then I would split them now.

Comment (benhoyt) [reply ops/testing.py:1517]: No, not really. I got the sense they wanted to decide soon, but I'm not sure how long that piece of string is...

I'd probably recommend not doing the rename now and leaving that for a future PR.

Comment (benhoyt) [reply ops/testing.py:1517]: (so that we can merge this work without blocking).

**Comment by IronCore864** [comment]
John has started a discussion in the juju development channel, now the preferred names are "application secrets" and "charm secrets", it seems it's not decided yet, I'll wait for a final conclusion before merging this one.

**Review by tonyandrewmeyer — APPROVED** [review]
LGTM!

**Review by benhoyt — APPROVED** [review]

**Pull request merged**
